src/mono/mono/mini/mini-runtime.c

```diff
@@ -4564,6 +4560,10 @@ mini_init (const char *filename, const char *runtime_version)
 	else
 		domain = mono_init_from_assembly (filename, filename);
 
+#ifdef ENABLE_PERFTRACING
+	ep_init ();
+#endif
```

Comment:
When adding ability to start a EP session using environment, there was a couple of calls that ends up checking GC policy and that is not initialized until mono_init_internal, the thread is not attached until that point as well, so if anything happens that will end up in runtime (like flushing file stream) that will assert. One option could be to split init of EP and do the potential setup of the environment based file session at a later point in time (after mono_init_internal), but regardless, if you can't create a EP session until later there won't be a way to get hold of events happening before that moment, so doing it early will only mean you could do some init that needs to be in place for other functions to work (like profiler) at a later time.

Reply:
I will probably revisit this later when we have diagnostic server up and running, maybe we could make the IPC streaming working during early stages of runtime init and just postpone the file stream env based to later (if used). Another alternative would be to try getting away from calling any runtime functions that depends on GC transition during init (always running in safe mode), but not sure if that is possible since some part depends on allocating runtime primitives like events/locks/threads, so then we probably need to implement a PAL or make parts of the runtime shim not depending on runtime. Yet another option would be to see if we could get a specific ep provider/session running during init that will be flushed into ep sessions established early (but after runtime has init). There is also work done in this PR, https://github.com/dotnet/runtime/pull/36720, that I will look at when porting diagnostic server.